source/slang/slang-ir-legalize-global-values.cpp

```diff
@@ -0,0 +1,233 @@
+#include "slang-ir-legalize-global-values.h"
+
+#include "slang-ir-clone.h"
+#include "slang-ir-util.h"
+
+namespace Slang
+{
+
+void GlobalInstInliningContextGeneric::inlineGlobalValuesAndRemoveIfUnused(IRModule* module)
+{
+    List<IRUse*> globalInstUsesToInline;
+    List<IRInst*> globalInstsToConsiderDeleting;
+
+    for (auto globalInst : module->getGlobalInsts())
+    {
+        if (isInlinableGlobalInst(globalInst))
+        {
+            for (auto use = globalInst->firstUse; use; use = use->nextUse)
+            {
+                if (getParentFunc(use->getUser()) != nullptr)
+                {
+                    globalInstUsesToInline.add(use);
+                }
+            }
+            globalInstsToConsiderDeleting.add(globalInst);
+        }
+    }
+
+    for (auto use : globalInstUsesToInline)
+    {
+        auto user = use->getUser();
+        IRBuilder builder(user);
+        builder.setInsertBefore(getOutsideASM(user));
+        IRCloneEnv cloneEnv;
+        auto val = maybeInlineGlobalValue(builder, use->getUser(), use->get(), cloneEnv);
+        if (val != use->get())
+            builder.replaceOperand(use, val);
+    }
+
+    // Since certain globals that appear in the IR are considered illegal for all targets,
+    // e.g. calls to functions, we delete globals which no longer have uses after inlining.
+    for (auto& globalInst : globalInstsToConsiderDeleting)
+    {
+        // TODO: Explain why tests/spirv/global-compute.slang fails if we don't exclude
```

Comment:
Todo in comment should be addressed.

Reply:
The issue was that it wasn't actually safe to remove the global inst, because I didn't check whether the inst _or any of its descendants_ had any usages.

For the mentioned global-compute test, there were some `SPIRVAsmOperandInst` descendants of the global inst that had usages left after inlining. (They didn't get inlined and so the inlined `SPIRVAsm` itself was using them.)

Is the fact that the `SPIRVAsmOperandInst`s didn't get inlined an issue we should fix?
If so I would like to file a new task for that, and do it in a later commit. (Right now it's not causing any problem.)